### PR TITLE
Make lr and momentum shared_scalars

### LINKED ISSUE #536: Learning rate and momentum as shared scalars
I'd like to ask if we could make the optimizer's learning rate (`lr`) and momentum (`momentum`) into shared scalars. This way we could change their values during training with `.set_value` using custom rules.

I could work on the PR as long as you guys don't have a reason for not to.

### PR DESCRIPTION
**Pull request by EderSantana**
Fix #536, make what the title says, some pep8 and add LrSetter callback

ToDo:
- [ ] Add a test to be used as LrSetter example

### CODE REVIEW

**Comment by fchollet** [comment]
- no PEP8-related changes, please.
- naming such as `LrSetter` should be avoided, because out of context it isn't clear what `lr` means (the `callbacks` module is not directly related to optimizers...). Prefer `LearningRateScheduler` or similar.
- I don't think learning rate scheduling should happen via a dictionary, instead if should probably take a lambda or function as argument, that would map an epoch to a learning rate
- usage should be explicitly restricted to optimizers that do have learning rates.

**Comment by EderSantana** [comment]
I can revert the pep8 changes and change the class name.

But what about this lambda function? The idea of keeping the learning rate fixed and changing at specific points (and keep it fixed for another while) is what is usually reported. Say I start with learning rate .01 and after 100 epochs change it to .0001, run another 100 epochs and finish training.

What do you mean by optimizers that have learning rates? Do you think we may have problems changing the learning rate of `adam`, `rmsprop` and such?

**Comment by fchollet** [comment]
> But what about this lambda function? The idea of keeping the learning rate fixed and changing at specific points (and keep it fixed for another while) is what is usually reported.

The most generic way to achieve this is via a function-based lookup of the current epoch. Dictionary-based lookup is a subset of function-based lookup, so it's essentially a more elegant / more generic way to do the same. 

> Say I start with learning rate .01 and after 100 epochs change it to .0001, run another 100 epochs and finish training.

Then use:
```python
LearningRateScheduler(schedule = lambda e: .01 if e < 100 else .0001)
```
Another more interesting possibility with a lambda: 
```python
LearningRateScheduler(schedule = lambda e: .1/(e + 1))
```
More interesting stuff also possible with full-on functions.

> What do you mean by optimizers that have learning rates? Do you think we may have problems changing the learning rate of adam, rmsprop and such?

With these methods, learning rate is implicit. You're not supposed to modify it during training.

**Comment by wuaalb** [comment]
I'd also much prefer (lambda) function for schedule.

Would it be possible to pass the name of the hyper-parameter that you wish to schedule to the callback's constructor?
This way a single class could work for lr, momentum, etc.

**Comment by EderSantana** [comment]
the problem is, only SGD got both momentum and lr. `momentum` got different names for other methods. So I only made the learning rate shared. 

I'm still not getting the lambda function, though. I'll do it since it already got 2 votes against 1. But the thing is, if you need a callback inside a callback, what would be the reason of the first one at all? Can you guys show me an example of what you need?

**Comment by wuaalb** [comment]
I was thinking something like this..

```python
class OptimizerScheduler(Callback):
    def __init__(self, schedule_func, param_name='lr'):
        super(OptimizerScheduler, self).__init__()
        self.schedule_func = schedule_func
        self.param_name = param_name

    def on_epoch_begin(self, epoch, logs={}):
        getattr(self.model.optimizer, self.param_name).set_value(self.schedule_func(epoch))
```

.. and if there's more than one hyper-parameter, you could add multiple callbacks.

**Comment by EderSantana** [comment]
PR here https://github.com/EderSantana/keras/blob/master/keras/callbacks.py
I accept it right away and everything gets merged together.

**Comment by EderSantana** [comment]
I didn't made all the possible parameters into shared scalars though. Do you think it would be necessary to change stuff like Adam's b's this way?
Otherwise, do a little check to say that only learning rates and momentum are modifiable.
